### PR TITLE
add address curd with restaurant relation

### PR DESCRIPTION
**Pull request by mosayed01**
https://www.postman.com/bold-comet-974062/workspace/beepbeep/overview
- add address curd and relation with restaurant
- get addresses
- get address by id
- get addresses in restaurant
- add address
- add addresses in restaurant 
- update address
- delete address

### CODE REVIEW

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/data/collection/RestaurantCollection.kt [hunk 1]
```diff
@@ -24,6 +25,9 @@ data class RestaurantCollection(
     val id: ObjectId = ObjectId()
     val isDeleted: Boolean = false
 
+    val addressIds: List<@Contextual Id<AddressCollection>> = emptyList()
```
Comment (NadaFeteiha) [inline service_restaurant/src/main/kotlin/org/thechance/service_restaurant/data/collection/RestaurantCollection.kt:28]: why type AddressCollection, not ObjectId or Id<String>?

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/entity/Restaurant.kt [hunk 2]
```diff
@@ -13,8 +12,8 @@ data class Restaurant(
     val phone: String? = null,
     val openingTime: String? = null,
     val closingTime: String? = null,
-    val isDeleted: Boolean = false
-//    val address: ObjectId,
+    val isDeleted: Boolean = false,
+    val addresses: List<String>? = null,
```
Comment (NadaFeteiha) [inline service_restaurant/src/main/kotlin/org/thechance/service_restaurant/entity/Restaurant.kt:16]: I think should be ' List < Address > ' not string

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/data/gateway/RestaurantGatewayImp.kt [hunk 3]
```diff
@@ -23,11 +25,71 @@ import org.thechance.service_restaurant.utils.isSuccessfullyUpdated
...
+                from = "addressCollection",
+                localField = RestaurantCollection::addressIds.name,
+                foreignField = "_id",
+                newAs = RestaurantCollection::addresses.name
+            ),
+            project(
+                RestaurantCollection::addresses
+            )
+        ).toList().firstOrNull()?.addresses?.filter { !it.isDeleted }?.toEntity() ?: emptyList()
+    }
```
Comment (NadaFeteiha) [inline service_restaurant/src/main/kotlin/org/thechance/service_restaurant/data/gateway/RestaurantGatewayImp.kt:90]: please check if the address is deleted inside aggregation.

**Review by NadaFeteiha — CHANGES_REQUESTED** [review]

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/data/collection/RestaurantCollection.kt (hunk 1 above)
Comment (mosayed01) [reply service_restaurant/src/main/kotlin/org/thechance/service_restaurant/data/collection/RestaurantCollection.kt:28]: It's already Id for AddressCollection

**Review by NadaFeteiha — APPROVED** [review]

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/entity/Restaurant.kt [hunk 4]
```diff
@@ -13,8 +12,8 @@ data class Restaurant(
     val phone: String? = null,
     val openingTime: String? = null,
     val closingTime: String? = null,
-    val isDeleted: Boolean = false
-//    val address: ObjectId,
+    val isDeleted: Boolean = false,
+    val addresses: List<Address>? = null,
 //    val meals: List<ObjectId>,
```
Comment (Asiasama710) [inline service_restaurant/src/main/kotlin/org/thechance/service_restaurant/entity/Restaurant.kt:17]: unused comment

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/usecase/address/CreateAddressUseCaseImpl.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import org.thechance.service_restaurant.api.usecases.CreateAddressUseCase
+import org.thechance.service_restaurant.entity.Address
+import org.thechance.service_restaurant.usecase.gateway.RestaurantGateway
+
+@Single
+class CreateAddressUseCaseImpl(
+    private val restaurantGateway: RestaurantGateway
+) : CreateAddressUseCase {
+    override suspend fun invoke(address: Address): Boolean {
+        return restaurantGateway.addAddress(address)
```
Comment (Asiasama710) [inline service_restaurant/src/main/kotlin/org/thechance/service_restaurant/usecase/address/CreateAddressUseCaseImpl.kt:13]: function in Gateway should be same name of usecase

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/usecase/address/GetAddressDetailsUseCaseImpl.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import org.thechance.service_restaurant.api.usecases.GetAddressDetailsUseCase
+import org.thechance.service_restaurant.entity.Address
+import org.thechance.service_restaurant.usecase.gateway.RestaurantGateway
+
+@Single
+class GetAddressDetailsUseCaseImpl(
+    private val restaurantGateway: RestaurantGateway
+): GetAddressDetailsUseCase {
+    override suspend fun invoke(addressId: String): Address {
+        return restaurantGateway.getAddress(addressId) ?: throw Throwable()
```
Comment (Asiasama710) [inline service_restaurant/src/main/kotlin/org/thechance/service_restaurant/usecase/address/GetAddressDetailsUseCaseImpl.kt:13]: function in Gateway should be same name of usecase

File: service_restaurant/src/main/kotlin/org/thechance/service_restaurant/entity/Restaurant.kt (hunk 4 above)
Comment (mosayed01) [reply service_restaurant/src/main/kotlin/org/thechance/service_restaurant/entity/Restaurant.kt:17]: deleted in develop branch

**Review by devsadeq — APPROVED** [review]
Looks good to me.

**Review by Asiasama710 — COMMENTED** [review]
well done  🥳 ,  but u should  solve conflict.

**Pull request merged**